### ISSUE TITLE
v5: Stack.Screen renders twice when switching to another Stack.Navigator

### ISSUE THREAD

**Issue by ChangeSomeCode** [react-navigation/react-navigation#8195]
**Current Behavior**

Hello!

I came across a situation where `Stack.Screen` renders twice when switching to another `Stack.Navigator`.

Using the example from the [documentation(auth-flow)](https://reactnavigation.org/docs/auth-flow/), I made a render of stacks by condition `isSignedIn === true`.

There are two components - `HomeStack` and `LoginStack`, which return different `Stack.Navigator`

```javascript
function HomeStack(props) {
  return (
    <Stack.Navigator>
      <Stack.Screen name="Home" component={HomeScreen} />
    </Stack.Navigator>
  );
}

function LoginStack({setIsSignedIn}) {
  return (
    <Stack.Navigator>
      <Stack.Screen name="SignIn">
        {(props) => <SignInScreen {...props} setIsSignedIn={setIsSignedIn} />}
      </Stack.Screen>

      <Stack.Screen name="SignUp">
        {(props) => <SignUpScreen {...props} setIsSignedIn={setIsSignedIn} />}
      </Stack.Screen>
    </Stack.Navigator>
  );
}
```

One of these stacks is rendered by condition inside the `NavigationContainer`:

```javascript
function App() {
  const [isSignedIn, setIsSignedIn] = useState(false);

  return (
    <NavigationContainer>
      {(isSignedIn) ? (
        <HomeStack />
      ) : (
        <LoginStack setIsSignedIn={setIsSignedIn} />
      )}
    </NavigationContainer>
  );
}
```

**Expected Behavior**

`HomeScreen` component should be rendered once.

**How to reproduce**

If we go from the initial route `SignIn` to the next `SignUp`, and then set the value for `isSignedIn = true`, app renders `HomeScreen` twice inside `HomeStack`.

If we set `isSignedIn = true` from the initial `SignIn` route, app renders `HomeScreen` once.

Also, if we put `Stack.Navigator` from `HomeStack` and `LoginStack` inside the `NavigationContainer`, then everything works fine.

```javascript
function App() {
  const [isSignedIn, setIsSignedIn] = useState(false);
  
  return (
    <NavigationContainer>
      {(isSignedIn) ? (
        <StackOne.Navigator initialRouteName="Home">
          <StackOne.Screen name="Home" component={HomeScreen} />
        </StackOne.Navigator>
      ) : (
        <StackTwo.Navigator initialRouteName="SignIn">
          <StackTwo.Screen name="SignIn">
            {(props) => <SignInScreen {...props} setIsSignedIn={setIsSignedIn} />}
          </StackTwo.Screen>

          <StackTwo.Screen name="SignUp">
            {(props) => <SignUpScreen {...props} setIsSignedIn={setIsSignedIn} />}
          </StackTwo.Screen>
        </StackTwo.Navigator>
      )}
    </NavigationContainer>
  );
}
```

example repo: https://github.com/ChangeSomeCode/react-navigation-render-twice-issue

change used `AppRenderTwice` or `AppWorksFine` in `index.js` for see what hapend in different `NavigationContainer` stacks.

**Your Environment**

| software                       | version |
| ------------------------------ | ------- |
| Android                        |
| @react-navigation/native       | 5.2.3
| @react-navigation/stack        | 5.2.18
| react-native-gesture-handler   | 1.6.1
| react-native-safe-area-context | 0.7.3
| react-native-screens           | 2.7.0
| react-native                   | 0.62.2
| expo                           | 
| node                           | v10.20.1
| npm                            | 6.14.4

**Comment by satya164** [comment]
Why does it matter?

**Comment by ChangeSomeCode** [comment]
because two instances are created.

of course we can use a working variant, but this is unexpected behavior, anyway.

**Comment by satya164** [comment]
Two renders don't create two instances of component, and most of the time re-renders don't matter. If a library re-renders something, it's probably an implementation detail and you shouldn't be caring about it.

In this case you need to use `React.memo` as mentioned in docs if you're concerned about re-renders https://reactnavigation.org/docs/hello-react-navigation#passing-additional-props

**Comment by ChangeSomeCode** [comment]
I checked the `props.navigation` state and the instances had a different route keys

**Comment by satya164** [comment]
You're talking about mount, not render. 2 mounts are indeed a bug.

**Comment by rumax** [comment]
I get the same behaviour with rendering twice but for a combination of Tabs and Stack, when stack navigator has tabs as a child screen. In this case child screen (tabs navigator) rendered twice with the same `params: undefined` but different `key`.

Any suggestion how to debug/check/solve it?

**Comment by ChangeSomeCode** [comment]
> You're talking about mount

yeah you right! my bad)

**Comment by ChangeSomeCode** [comment]
thanks!

**Comment by JulianKingman** [comment]
Wow, that was found and fixed fast! Will this be in bottom-tabs 5.3.5?

**Comment by rosnaib11** [comment]
Exactly same issue to me. I'm using React.createRef https://reactnavigation.org/docs/navigating-without-navigation-prop/.  It rendering twice when going from sign-in to home screen.

**Comment by rodriguezmanu** [comment]
+1

**Comment by marsicdev** [comment]
+1 same here still

**Comment by zhahaoyu** [comment]
This is happening with my app recently as well. Any regression?

**Comment by mariansam** [comment]
This is happening again. I'm using a combination of auth-flow, bottom-tabs and stack-navigator.

I've added a piece of code like this to every screen:
```ts
useEffect(() => {
    console.log('screen mounted');

    return () => {
        console.log('screen will unmount');
    }
}, []);
```

And tried changing the settings of the navigators a bit and I'm pretty sure this happens only after the user logs in, meaning the bottom-tabs-navigator gets opened for the first time. So this happens only for the screen specified in the `Navigator`'s `initialRouteName`.

**Comment by Jackyaung** [comment]
i got the same problem, how you solved it ?

**Comment by mariansam** [comment]
@Jackyaung I haven't. You can add a piece of code like this:
```ts
useEffect(() => {
    const hack = setTimeout(doStuff, 10);
    return () => {
        clearTimeout(hack);
    };
}, []);
```
to the affected screen. You can do all your initialization in the `doStuff` function.

**Comment by AsmaTidafi** [comment]
same here, any solution?

**Comment by umardevid** [comment]
+1 same here still

**Comment by therunbuddy** [comment]
Did anyone manage to solve this? Its a very annoying bug that when I switch the the insecure stack after logging out the screen will flash twice i.e. re-render

**Comment by pbroeker** [comment]
+1. Anyone opened a new issue for the development team regarding this?

**Comment by therunbuddy** [comment]
Nope I dont think so. Would you be able to do it?

**Comment by allemanfredi** [comment]
This is a solution that is not very clean but that can work:

```js
import { Mutex } from 'async-mutex'

class SMutex {
  static getInstance() {
    if (!SMutex.instance) {
      SMutex.instance = new Mutex()
    }
    return SMutex.instance
  }
}

const MyComponent = () => {
  useEffect(() => {
    const doStuff = async () => {
      if (SMutex.getInstance().isLocked()) {
        return
      }
      const release = await SMutex.getInstance().acquire()

      try {
       ...
      } catch (_err) {
       ...
      } finally {
        release()
      }
    }

    doStuff()
  }, [])

  return ...
}
```

**Comment by bimusiek** [comment]
If someone comes here, the issue is using function to return component.
https://reactnavigation.org/docs/troubleshooting/#screens-are-unmountingremounting-during-navigation
